cmd/pj-rehearse-metrics/pj-rehearse-metrics.go

```diff
@@ -0,0 +1,172 @@
+	if _, err := os.Stat(localMetricsPath); err == nil {
+		return
+	}
+
+	reader, err := s.bucket.Object(metricsPath).NewReader(s.ctx)
+	if err != nil {
+		return
+	}
+	data, err := ioutil.ReadAll(reader)
+	_ = reader.Close()
+	if err != nil {
+		fmt.Printf("could not read rehearse metrics err=%v\n", err)
+		return
+	}
+
+	if err := os.MkdirAll(prDir, 0755); err != nil {
+		fmt.Printf("could not create a directory for PR err=%v\n", err)
+		return
+	}
+
+	if err := ioutil.WriteFile(localMetricsPath, data, 0644); err != nil {
+		fmt.Printf("could not write metrics to %s err=%v\n", localMetricsPath, err)
+		return
+	}
+}
+
+func (s *scraper) scrapePr(prefix string) {
+	prNumber := path.Base(prefix)
+	jobPrefix := path.Join(prefix, rehearseJobName)
+	jobPrefix = fmt.Sprintf("%s/", jobPrefix) // Needed because path.Join() removes trailing slash
+
+	runs := s.bucket.Objects(s.ctx, &storage.Query{Prefix: jobPrefix, Delimiter: "/"})
+	for {
+		attrs, err := runs.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			fmt.Printf("Next() failed err=%v\n", err)
+			break
+		}
+
+		// Leaf (a file), not a prefix (a dir)
+		if attrs.Name != "" {
+			continue
+		}
+
+		s.scrapeRun(attrs.Prefix, prNumber)
+	}
+}
+
+type options struct {
+	cacheDir string
+}
+
+func gatherOptions() options {
+	o := options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+
+	fs.StringVar(&o.cacheDir, "cache-dir", "", "Path to a directory where scraped metrics data will be saved. If not provided, cache will not be used.")
+
+	_ = fs.Parse(os.Args[1:])
+	return o
+}
+
+func main() {
+	o := gatherOptions()
+	var err error
+
+	if o.cacheDir == "" {
+		o.cacheDir, err = ioutil.TempDir("", "")
+		if err != nil {
+			_, _ = fmt.Fprintf(os.Stderr, "Failed to create a temporary directory: %v", err)
+			os.Exit(1)
+		}
+		defer os.RemoveAll(o.cacheDir)
+	}
+
+	ctx := context.Background()
+
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		_, _ = fmt.Fprintf(os.Stderr, "Failed to create client: %v", err)
+		os.Exit(1)
+	}
+
+	bucket := client.Bucket(originCiBucket)
+	pullRequests := bucket.Objects(ctx, &storage.Query{
+		Prefix:    "pr-logs/pull/openshift_release/",
+		Delimiter: "/",
+	})
+
+	scraper := scraper{ctx: ctx, bucket: bucket, cacheDir: o.cacheDir}
+	prPrefixes := make(chan string, 100)
+	progress := make(chan string, 100)
+
+	wg := sync.WaitGroup{}
+	wg.Add(20)
+
+	for w := 0; w < 20; w++ {
+		go func() {
+			for prefix := range prPrefixes {
+				scraper.scrapePr(prefix)
+				progress <- prefix
+			}
+			wg.Done()
+		}()
+	}
+
+	go func() {
+		for {
+			attrs, err := pullRequests.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				fmt.Printf("Next() failed: err=%v\n", err)
+				break
+			}
+
+			if attrs.Name != "" {
+				continue
+			}
+
+			prPrefixes <- attrs.Prefix
+		}
+		close(prPrefixes)
+		wg.Wait()
+		close(progress)
+	}()
+
+	counter := 0
+	for done := range progress {
```

Comment:
its a go idiom to use the `for{ select {} }` to handle also other signal channels to gracefully terminate.  
Also you can check if the channel closed. Like:
```go
 	for done, ok := range progress {
if !ok {
// we are done
os.Exit(0) 
}
}
```

Reply:
About the "check if the channel is closed" part, are you sure? The [spec](https://golang.org/ref/spec#For_range) says _If the range expression is a channel, at most one iteration variable is permitted_. It also does not make sense, I understand then when the channel is closed and all values are exhausted, the iteration simply ends...

Reply:
Yeah unless you want to cancel in the middle you don't need the `range` and this will stop looping when the channel is closed